app/models/asp/payment_request.rb

```diff
@@ -86,6 +86,8 @@ def unpaid_reason
     private
 
     def single_active_payment_request_per_pfmp
+      return if pfmp.nil?
```

Comment:
Une payment_request peut exister sans PFMP associée?

Reply:
c'est déclenché par le 

```ruby
 describe "validations" do
    it { is_expected.to validate_length_of(:asp_payment_requests).is_at_most(7000) }
  end
```

donc je pense que ça créé un record qui est pas persisté (donc PFMP en mémoire + PR pas encore persistée non plus) et du coup ça plante mais je vais re-regarder parce qu'en vrai je suis 99% sûr de ça et le 1% qui manque me dérange

Reply:
Normalement non, mais je vois effectivement que la colonne n'est pas obligatoire dans le schéma : https://github.com/betagouv/aplypro/blob/dontmerge/payments/db/schema.rb/#L61

Reply:
Mais je pense que la relation avec une PFMP devrait être obligatoire.

Reply:
OK je crois que j'ai compris grâce à cette issue : https://github.com/thoughtbot/shoulda-matchers/issues/1209#issuecomment-494547331

En fait Shoulda Matchers vérifie que la requête est bien invalide avec > 7000 requêtes de paiement, sauf qu'il fait un `klass.new` (https://github.com/thoughtbot/shoulda-matchers/blob/main/lib/shoulda/matchers/active_model/validate_length_of_matcher.rb#L492-L500) sur la classe associée, ce qui créé une PR sans `ASP::Request` ou `Pfmp`. En essayant de valider que ça marche pas (has_many trigger la validation des records associés), le callback de validation qui dépend de la présence d'une PFMP foire mais c'est un dommage collatéral.

Vais voir pour une solution élégante

Reply:
alors c'est un peu la misera pour créer 7000 PR (une factory très lourde vu qu'elle est en bout de chaîne de notre logique métier (Établissement -> Classe/MEF -> Schooling -> Élève -> PFMP -> PR) donc je vous propose soit de tej la validation soit de garder le `return if pfmp.nil?` + une contrainte de `not null` pour `pr.pfmp_id`. 

Donc rendre obligatoire la présence de la PFMP (ce qui est déjà le cas, juste pas dans la DB), mais faire en sorte que le garde puisse gérer ces états in-memory/non-persisted – d'ailleurs ces contraintes de présence ne doivent être honorés qu'au moment de la persistence des données. C.f les RIBs où les validations pour `iban` et `bic` checkent la valeur d'abord au cas où elles soient dans l'ordre de validation de ActiveRecord (`iban` et `rib` sont toutes les deux `not null` en DB et dans le modèle).

```rb

  before_validation do
    self.iban = Rib.normalize_value_for(:iban, iban) unless iban.nil?
    self.bic = Rib.normalize_value_for(:bic, bic) unless bic.nil?
  end
```

Reply:
Si on peut pas y couper à cette ligne dans ce cas là je serais partant d'ajouter un check d'environnement du style `if Rails.env.test? && pfmp.nil?` avec un gros commentaire qui explique un peu. `XXX: shoulda matcher quirk, blabla`

Reply:
en y repensant c'est pas vraiment une question de test, c'est juste que l'ORM ne te garantit pas forcément un objet complet parce qu'il est peut-être en train d'être créé, comme l'exemple des BICs au-dessus. Ton validateur check une propriété de la PFMP, mais s'il n'y a pas de PFMP alors il ne peut rien faire (P => Q, si P est faux tu peux rien faire de Q), donc je pense que c'est légitime de garder le garde, comme pour l'exemple des RIBs au dessus.

On est quand même assurés que la PFMP sera là parce qu'au moment de faire la persistence le `belongs_to :pfmp` va vérifier que l'objet est bien là + le `null: false` va s'assurer qu'il y a bien une valeur en DB et le `foreign_key: true` (dans le `schema.db`) va s'assurer qu'il existe bien une PFMP correspondante avec cet index, mais jusqu'à ce qu'on arrive au layer de persistence (save/create) on peut pas encore assumer ça.

Reply:
d'ailleurs les exemples du guide Rails vont dans ce sens aussi : 

https://guides.rubyonrails.org/active_record_validations.html#custom-methods

Reply:
d'ailleurs je me demande si cette validation n'est pas dans le mauvais sens, en terme de contexte/connaissance des objets la PFMP est plus encline à valider ses `payment_requests` plutôt que l'autre sens non ?